### PR TITLE
feat(feature-flag): dashboard on usage tab

### PR DESCRIPTION
**Pull request by EDsCODE**
## Problem

- usage tab is minimal right now
- attempt [2](https://github.com/PostHog/posthog/pull/14383)

## Changes
- auto create a dashboard and associate it with the feature flag to be shown in usage tab

Screen Shot 2023-02-23 at 10 13 35 AM

👉 *Stay up-to-date with [PostHog coding conventions](https://posthog.com/docs/contribute/coding-conventions) for a smoother review.*

### CODE REVIEW

**Comment by EDsCODE** [comment]
TODO: 
- [x] make sure it's backwards compatible (usage tabs that don't have connected dashboards work)

File: posthog/api/test/__snapshots__/test_decide.ambr [hunk 1]
```diff
@@ -171,7 +171,8 @@
          "posthog_featureflag"."active",
          "posthog_featureflag"."rollback_conditions",
          "posthog_featureflag"."performed_rollback",
-         "posthog_featureflag"."ensure_experience_continuity"
+         "posthog_featureflag"."ensure_experience_continuity",
+         "posthog_featureflag"."usage_dashboard_id"
```
Comment (neilkakkar) [inline posthog/api/test/__snapshots__/test_decide.ambr:175]: Got to be careful of this perf wise! I'd prefer if we didn't select it at all, but ok with this as long as no joins come into the picture 👀

**Comment by neilkakkar** [comment]
How will old flags work? Will we create new dashboard on click of usage tab?

**Comment by neilkakkar** [comment]
LGTM so far!

**Comment by liyiy** [comment]
👍

**Comment by liyiy** [comment]
> How will old flags work? Will we create new dashboard on click of usage tab?

old flags don't have dashboards on them so far but I think ^ is a good idea

File: posthog/api/test/__snapshots__/test_decide.ambr (hunk 1 above)
Comment (EDsCODE) [reply posthog/api/test/__snapshots__/test_decide.ambr:175]: Shouldn't need joins. The component on the frontend just needs the id

**Comment by EDsCODE** [comment]
Decided to add a tiny prompt to generate a dashboard. This is to account for the edge case where someone deletes the existing dashboard that's connected. I initially tried some work around adding a permission that would only allow our system to edit/delete flags but it's a bit convoluted because the permissioning feature is paywalled

Screen Shot 2023-02-28 at 2 40 53 PM

File: ee/clickhouse/views/test/__snapshots__/test_clickhouse_experiment_secondary_results.ambr [hunk 2]
```diff
@@ -1,4 +1,52 @@
 # name: ClickhouseTestExperimentSecondaryResults.test_basic_secondary_metric_results
+  '
+  /* user_id:46 celery:posthog.celery.sync_insight_caching_state */
```
Comment (neilkakkar) [inline ee/clickhouse/views/test/__snapshots__/test_clickhouse_experiment_secondary_results.ambr:3]: Do you know why are these snapshots being added here?

**Review by neilkakkar — APPROVED** [review]
Not sure what's going on with snapshots, but otherwise LGTM!

(I'd say investigate the experiment snapshots, those seem sus, maybe because of dashboard being added to the experiment flags automatically on creation?)

**Comment by annikaschmid** [comment]
This looks good to me so far! Some ideas around next steps:

1. Should we have a graph with a breakdown by flag variant, including false? This might be interesting in combination with the rollout, so users can easily see how the rollout has developed over time
2. As far as I can see, we don't support the dashboard settings that allow filtering and setting a custom time frame. I can see use cases for both of those, e.g. when the flag has been rolled out for more than 30 days, or I want to know whether a specific company was exposed to the flag already

If I understand correctly, this creates a dashboard, right? What happens if I add/remove graphs when I navigate to the dashboard? Will those be reflected on the usage page here?

**Comment by EDsCODE** [comment]
> Should we have a graph with a breakdown by flag variant, including false? This might be interesting in combination with the rollout, so users can easily see how the rollout has developed over time

One of them is a breakdown! I will relabel

> As far as I can see, we don't support the dashboard settings that allow filtering and setting a custom time frame. I can see use cases for both of those, e.g. when the flag has been rolled out for more than 30 days, or I want to know whether a specific company was exposed to the flag already

Fair, let me add these back and see how they fit

> If I understand correctly, this creates a dashboard, right? What happens if I add/remove graphs when I navigate to the dashboard? Will those be reflected on the usage page here?

Yeah, you can hack around this and add stuff to dashboard. I was messing with permissions to try to restrict this but i don't think it's a dealbreaker. If we find that users are heavily editing these dashboards, then we know there's opportunity to make a flexible dashboarding in flags experience

File: ee/clickhouse/views/test/__snapshots__/test_clickhouse_experiment_secondary_results.ambr (hunk 2 above)
Comment (EDsCODE) [reply ee/clickhouse/views/test/__snapshots__/test_clickhouse_experiment_secondary_results.ambr:3]: Now that creating feature flags will create dashboards, the [post_save hook](https://github.com/PostHog/posthog/blob/master/posthog/models/insight_caching_state.py#L58) will be called for dashboard tiles

File: posthog/api/feature_flag.py [hunk 3]
```diff
@@ -239,6 +234,23 @@ def _update_filters(self, validated_data):
...
 
+def _create_usage_dashboard(feature_flag: FeatureFlag, user):
+    from posthog.helpers.dashboard_templates import create_feature_flag_dashboard
+    from posthog.models.dashboard import Dashboard
+
+    usage_dashboard = Dashboard.objects.create(
+        name="Auto-Generated Dashboard: " + feature_flag.key + " Usage",
+        description="This dashboard was autogenerated when the feature flag with key ("
+        + feature_flag.key
+        + ") was created",
```
[the remark this answers is not in the document]
Comment (EDsCODE) [reply posthog/api/feature_flag.py:245]: oops

**Pull request merged**